### PR TITLE
Disputes: Implement evidence period

### PR DESCRIPTION
**Pull request by facuspagnuolo**
This PR aims to implement the evidence submission period for disputes right before the first draft

### CODE REVIEW

File: contracts/arbitration/IArbitrable.sol [hunk 1]
```diff
@@ -1,8 +1,21 @@
...
 
+import "./IArbitrator.sol";
+
 
 interface IArbitrable {
-    // bytes4 constant INTERFACE_ID = 0x311a6c56;
+    // bytes4 constant INTERFACE_ID = 0x88f3ee69;
+
+    event Ruled(IArbitrator indexed arbitrator, uint256 indexed disputeId, uint256 ruling);
+    event EvidenceSubmitted(uint256 indexed disputeId, bytes evidence, bool finished);
```
Comment (bingen) [inline contracts/arbitration/IArbitrable.sol]: Can we add a reference to who submits the evidence?

File: contracts/court/AragonCourt.sol [hunk 2]
```diff
@@ -97,7 +99,18 @@ contract AragonCourt is Controller, IArbitrator {
...
     /**
-    * @dev Execute the arbitrable associated to a dispute based on its final ruling
+    * @notice Close the evidence period of dispute #`_disputeId`
+    * @param _disputeId Identification number of the dispute to close its evidence submitting period
+    */
+    function closeEvidencePeriod(uint256 _disputeId) external {
+        IDisputeManager disputeManager = IDisputeManager(_getDisputeManager());
+        (IArbitrable subject,,,,,) = disputeManager.getDispute(_disputeId);
+        require(subject == IArbitrable(msg.sender), ERROR_SENDER_NOT_DISPUTE_SUBJECT);
+        disputeManager.closeEvidencePeriod(_disputeId);
```
Comment (bingen) [inline contracts/court/AragonCourt.sol]: Maybe we can send `msg.sender` to `disputeManager.closeEvidencePeriod` and do the check inside. This way we save 2 calls to `disputeManager`.

File: contracts/court/config/CourtConfig.sol [hunk 3]
```diff
@@ -164,17 +166,19 @@ contract CourtConfig is IConfig, CourtConfigData {
...
         internal
     {
-        // If the current term is not zero, changes must be scheduled at least 2 terms in the future.
-        // This way we can ensure that disputes scheduled for the next term won't have their config changed.
-        require(_currentTermId == 0 || _fromTermId > _currentTermId + 1, ERROR_TOO_OLD_TERM);
+        // If the current term is not zero, changes must be scheduled at least after the next evidence period.
+        // This way we can ensure that disputes already scheduled won't have their config changed.
+        uint64 evidenceTerms = configs[configIdByTerm[_currentTermId]].disputes.evidenceTerms;
+        // No need for SafeMath: round state durations are safely capped
+        require(_currentTermId == 0 || _fromTermId > _currentTermId + evidenceTerms, ERROR_TOO_OLD_TERM);
```
Comment (bingen) [inline contracts/court/config/CourtConfig.sol]: If we are going to use creation term for dispute configs we don't need this check.

File: contracts/disputes/DisputeManager.sol [hunk 4]
```diff
@@ -181,26 +169,27 @@ contract DisputeManager is ControlledRecoverable, ICRVotingOwner, IDisputeManage
...
+    * @notice Create a dispute over `_subject` with `_possibleRulings` possible rulings
     * @param _subject Arbitrable instance creating the dispute
     * @param _possibleRulings Number of possible rulings allowed for the drafted jurors to vote on the dispute
     * @param _metadata Optional metadata that can be used to provide additional information on the dispute to be created
     * @return Dispute identification number
     */
     function createDispute(IArbitrable _subject, uint8 _possibleRulings, bytes calldata _metadata) external onlyController returns (uint256) {
         uint64 termId = _ensureCurrentTerm();
-        ISubscriptions subscriptions = _subscriptions();
-        require(subscriptions.isUpToDate(address(msg.sender)), ERROR_SUBSCRIPTION_NOT_PAID);
```
Comment (bingen) [inline contracts/disputes/DisputeManager.sol]: Why are we removing this?

File: contracts/disputes/DisputeManager.sol [hunk 5]
```diff
@@ -212,6 +201,26 @@ contract DisputeManager is ControlledRecoverable, ICRVotingOwner, IDisputeManage
         return disputeId;
     }
 
+    /**
+    * @notice Close the evidence period of dispute #`_disputeId`
+    * @param _disputeId Identification number of the dispute to close its evidence submitting period
+    */
+    function closeEvidencePeriod(uint256 _disputeId) external onlyController {
+        uint256 roundId = 0;
+        _checkRoundExists(_disputeId, roundId);
```
Comment (bingen) [inline contracts/disputes/DisputeManager.sol]: You know I like it better here, but why isn't this as a modifier?

File: contracts/arbitration/IArbitrable.sol (hunk 1 above)
Comment (facuspagnuolo) [reply contracts/arbitration/IArbitrable.sol]: Sure!

File: contracts/court/AragonCourt.sol (hunk 2 above)
Comment (facuspagnuolo) [reply contracts/court/AragonCourt.sol]: Yes, I did like you say at first, but then rolled back to this way due to size limits. Will try again.

File: contracts/court/config/CourtConfig.sol (hunk 3 above)
Comment (facuspagnuolo) [reply contracts/court/config/CourtConfig.sol]: True, good catch!

File: contracts/disputes/DisputeManager.sol (hunk 4 above)
Comment (facuspagnuolo) [reply contracts/disputes/DisputeManager.sol]: Cause it was already validated in the entry point (`AragonCourt`) :)

File: contracts/disputes/DisputeManager.sol (hunk 5 above)
Comment (facuspagnuolo) [reply contracts/disputes/DisputeManager.sol]: I know, simply to reuse the `roundId` variable, it is redundant tho

**Pull request merged**